Backend plugin debug mode improvements

xnyo opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana-plugin-sdk-go/blob/master/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

4. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the master branch.

-->

**What this PR does / why we need it**:

This PR improves the backend debug mode, allowing to:

- Automatically turning on debug mode when running in debug mode from GoLand
- Turn on debug mode manually by launching the plugin with a new `-debug` CLI flag (along with `-standalone`)
- Extends the `standalone.txt` definition by adding a PID at the end, which can be used by the dummy plugin locator to detect if the debug plugin is running and turn itself on/off depending on that
- Graceful shutdown of the standalone GRPC server, with `standalone.txt` file cleanup
- This PR also introduces some refactoring around standalone plugin management

For this to work, the plugin must be built with this new version of the SDK.


<!--
**Which issue(s) this PR fixes**:



* Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"


Fixes #

-->
**Special notes for your reviewer**:

Tested on Linux and Mac OS X with [datasource-http-backend](https://github.com/grafana/grafana-plugin-examples/tree/main/examples/datasource-http-backend) plugin example, adding the following to `go.mod` to compile against the SDK:

```
replace github.com/grafana/grafana-plugin-sdk-go => ../../../grafana-plugin-sdk-go
```

Where `../../../grafana-plugin-sdk-go` is the relative path to this repo checked out to this PR's branch

## GoLand setup

![immagine](https://user-images.githubusercontent.com/16373015/221838011-fedf2d01-3642-416e-b769-4afca36c2035.png)

Then run in debug mode.
To make it work also when running _without_ debug mode, add `-debug` to "Program arguments" (along with `-standalone`)

## Visual Studio Code setup

Put the following in `.vscode/launch.json`:

```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Standalone debug mode",
            "type": "go",
            "request": "launch",
            "mode": "debug",
            "program": "${workspaceFolder}/pkg",
            "env": {},
            "args": [
                "-standalone"
            ]
        }
    ]
}
``` 

Then press F5 to run the configuration. 

## Manual setup (/other IDEs)

```
$ mage build:debug
$ cd dist
$ ./gpx_PLUGIN_ID_PLATFORM_ARCH -standalone -debug
```

## Notes

You will have to restart Grafana the first time you enable debug mode, but once the debug plugin is detected, the plugin will be "reloaded" automatically if you re-compile it and re-run the standalone executable.

xnyo committed:
Add support for debug mode when running from GoLand

wbrowne commented on internal/standalone/standalone.go:
```diff
@@ -38,8 +38,12 @@ func GetInfo(id string) (Args, error) {
+	// When debugging, write the file in `dist`
+	// VsCode names the file "__debug_bin"
+	vsCodeDebug := strings.HasPrefix(filepath.Base(ex), "__debug_bin")
+	// GoLand places it in: /tmp/GoLand/___XXgo_build_github_com_PACKAGENAME_pkg
```
Do we know if this is consistent across platforms?

xnyo replied:
On Mac OS I got:
```
/usr/local/go/bin/go build -o /private/var/folders/lx/k8zcwy851kzdrxrzjwsv34gr0000gn/T/GoLand/___go_build_PACKAGENAME
```

So it should match anyways. Not sure about Windows, I will test shortly.

xnyo committed:
Add support for -debug flag to explicitly run the plugin in debug mode

xnyo committed:
Fix plugin killing itself when running with -debug flag

wbrowne commented on internal/standalone/standalone.go:
```diff
@@ -152,9 +158,14 @@ func findAndKillCurrentPlugin(dir string) {
 	for _, txt := range strings.Fields(string(out)) {
 		pid, err := strconv.Atoi(txt)
 		if err == nil {
+			// Do not kill ourselves!
```
```suggestion
			// Do not kill the plugin process
```

andresmgot commented:
Tested with vscode and still working as expected so :+1: :smile:

xnyo committed:
Update comments as per PR review

xnyo committed:
Update comment

xnyo committed:
Improve debug mode plugin management

- Store PID in standalone file, to allow Grafana to detect if a debug
  plugin is running or not. This allows Grafana to start the plugin with
  the default Hashicorp mode if the debug plugin isn't running anymore.
- Graceful shutdown of GRPC server, with standalone file cleanup

Note: needs to be tested on Mac OS X and Windows

xnyo commented:
I have added more improvements to the debug mode to allow to automatically switch off debug mode in Grafana.

I have converted the PR back to draft until I fully test it works properly on Mac and Windows.

xnyo committed:
Fix comment indentation

andresmgot reviewed (commented):
mmh, this is now not working for me. I am trying it with [this external plugin (ADX)](https://github.com/grafana/azure-data-explorer-datasource). When launching the debugger for the plugin, I get in the Grafana logs:

```
ERROR[03-01|12:18:07] Failed to restart plugin                 logger=plugin.grafana-azure-data-explorer-datasource error="Unsupported plugin protocol \"netrpc\". Supported: [grpc]"
```

And if after that, I try to restart, I get in the debugger terminal:

```
{"@level":"info","@message":"Profiler","@timestamp":"2023-03-01T12:18:18.694242+01:00","enabled":false}
{"@level":"info","@message":"Profiler","@timestamp":"2023-03-01T12:18:18.694304+01:00","enabled":false}
{"@level":"info","@message":"Creating standalone address file","@timestamp":"2023-03-01T12:18:18.694649+01:00"}
error running pgrep: exit status 1 (gpx_adx)
{"@level":"debug","@message":"Standalone plugin server","@timestamp":"2023-03-01T12:18:18.704547+01:00","capabilities":["diagnostics","resources","data","stream"]}
```

I suspect those are two different issues? (e.g. the second is probably due to the fact that the debug process does not have the same name than the `"executable"` specified in the plugin.json.

andresmgot commented:
> Hmm yeah a bit odd indeed. I will take a better look. Was this the same plugin that you used to test the previous version of the PR?

yes, I can try with some example plugin to see if I get the same behavior (but for ADX, it's working if I use the code from `main` so some change in this PR is affecting it).

xnyo commented:
> > Hmm yeah a bit odd indeed. I will take a better look. Was this the same plugin that you used to test the previous version of the PR?
> 
> yes, I can try with some example plugin to see if I get the same behavior (but for ADX, it's working if I use the code from `main` so some change in this PR is affecting it).

I tried running the adx plugin in debug mode and it's starting fine for me.

When you start the plugin in debug mode, a file called `standalone.txt` should be created in the plugin's dist folder. Can you send its content here please so I can take a better look? Thanks!

xnyo committed:
Split standalone address and pid in two different files
Fixes the dummy locator server not reading the address properly if it's been compiled with the old version SDK, but the standalone.txt comes from the new version of the SDK

xnyo commented:
also, PID has been added so that the dummy locator can turn itself off (and the plugin starts in normal mode) if standalone.txt exists but the process crashed. We check if the process exists or not using the PID. Before, you had to delete standalone.txt or the plugin would not work correctly, as it would attempt to connect to a GRPC server that's not running.

andresmgot reviewed (approved):
Now it works as before :+1: The only issue is that it still stops working if I stop the debugger process in the plugin. It's a minor thing since executing `mage reloadPlugin` or restarting the grafana server makes it work again but I am not sure if this PR was meant to fix that too. Anyway, my +1 to merge this as it is.

xnyo commented:
Thanks for the review @andresmgot !

> The only issue is that it still stops working if I stop the debugger process in the plugin. It's a minor thing since executing `mage reloadPlugin` or restarting the grafana server makes it work again but I am not sure if this PR was meant to fix that too. Anyway, my +1 to merge this as it is.

Yeah exactly, the PR was not meant to fix that. Unfortunately there's no reliable way to do to hat. Maybe we can kill the dummy locator when exiting the debugger process (which will reload the plugin in Grafana), but I don't think it'll be a very robust solution.

I will try that later and leave it for a follow-up PR :smile:

xnyo committed:
Merge branch 'main' into giuseppe/goland-debug

xnyo committed:
Fix standalone.txt not being cleaned

xnyo committed:
Kill dummy plugin locator when stopping debug plugin

xnyo committed:
Ensure plugin is running before starting dummy plugin locator

xnyo committed:
Fix GoLand debug mode when using non-standard config name, add some tests

xnyo commented:
@wbrowne this should now handle non-standard GoLand run configs (the issue you had last week). Please can you test and let me know if it works for you? Thanks :)

xnyo committed:
Merge branch 'main' into giuseppe/goland-debug

xnyo committed:
Fix goimports

wbrowne reviewed (approved):
LGTM
